lib/commands/web.js

```diff
@@ -834,6 +838,68 @@ const extensions = {
     }
   },
 
+  /**
+   * @this {XCUITestDriver}
+   * @returns {string} The base url which could be used to access WDA HTTP endpoints
+   * FROM THE SAME DEVICE where WDA is running
+   */
+  getWdaLocalhostRoot() {
+    const remotePort =
+      (this.opts.wdaRemotePort
+      ?? this.wda?.url?.port
+      ?? this.opts.wdaLocalPort)
+      || 8100;
+    return `http://127.0.0.1:${remotePort}`;
+  },
+
+  /**
+   * Calibrates web to real coordinates translation.
+   * This API can only be called from Safari web context.
+   * It must load a custom page to the browser, and then restore
+   * the original one, so don't call it if you can potentially
+   * lose the current web app state.
+   * The outcome of this API is then used in nativeWebTap mode.
+   * The returned value could also be used to manually transform web coordinates
+   * to real devices ones in client scripts.
+   *
+   * @this {XCUITestDriver}
+   * @returns {Promise<{offset: import('../types').Delta}>}
+   */
+  async mobileCalibrateWebToRealCoordinatesTranslation() {
+    if (!this.isWebContext()) {
+      throw new errors.NotImplementedError('This API can only be called from a web context');
+    }
+
+    const currentUrl = await this.getUrl();
+    await this.setUrl(`${this.getWdaLocalhostRoot()}/calibrate`);
+    const {width, height} = await this.getWindowRect();
+    const [centerX, centerY] = [Math.trunc(width / 2), Math.trunc(height / 2)];
+    await this.mobileTap(centerX, centerY);
+    const errorPrefix = 'Cannot determine web view coordinates offset. Are you in Safari context?';
+    let webClickCoordinates;
+    let title;
+    try {
+      title = await this.title();
+      this.log.debug(JSON.stringify(title));
+      webClickCoordinates = JSON.parse(title);
+    } catch (e) {
+      throw new Error(`${errorPrefix} Original error: ${e.message}`);
```

Comment:
I got the below error once in my testing. Just FYI.

That was only once after some find element -> click actions in my testing. (I modified `this.log.debug(JSON.stringify(title));` as `this.log.debug(JSON.stringify(title.x));` then)

```
[HTTP] --> POST /session/311ea990-ff4d-407f-a69b-5afac9c165e8/execute/sync
[HTTP] {"script":"mobile: calibrateWebToRealCoordinatesTranslation","args":[]}
[debug] [XCUITestDriver@e03c (311ea990)] Calling AppiumDriver.execute() with args: ["mobile: calibrateWebToRealCoordinatesTranslation",[],"311ea990-ff4d-407f-a69b-5afac9c165e8"]
[debug] [XCUITestDriver@e03c (311ea990)] Executing command 'execute'
[debug] [RemoteDebugger] Sending javascript command: 'window.location.href'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 80): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 80): '"safari-web-extension://6CACAC31-0394-4779-9D1B-B7AA651A9E32/_generated_background_page.html"'
[debug] [RemoteDebugger] Sending to Web Inspector took 65ms
[debug] [XCUITestDriver@e03c (311ea990)] Attempting to set url 'http://127.0.0.1:8200/calibrate'
[debug] [RemoteDebugger] Navigating to new URL: 'http://127.0.0.1:8200/calibrate'
[debug] [RemoteDebugger] Waiting for frame navigated message...
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 82): 'Page.navigate'
[debug] [RemoteDebugger] Received data response from send (id: 82): '{}'
[debug] [RemoteDebugger] Sending to Web Inspector took 30ms
[debug] [RemoteDebugger] Frame navigated in 502ms from: 500ms timeout
[debug] [RemoteDebugger] Waiting for dom...
[debug] [RemoteDebugger] Page loaded, verifying whether ready
[debug] [RemoteDebugger] Checking document readyState
[debug] [RemoteDebugger] Sending javascript command: 'document.readyState;'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 84): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 84): '"complete"'
[debug] [RemoteDebugger] Sending to Web Inspector took 14ms
[debug] [RemoteDebugger] Document readyState is 'complete'
[debug] [RemoteDebugger] Page is ready
[debug] [RemoteDebugger] Page load completed in 517ms
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 86): 'Console.enable'
[debug] [RemoteDebugger] Received data response from send (id: 86): '{}'
[debug] [RemoteDebugger] Sending to Web Inspector took 14ms
[debug] [RemoteDebugger] Executing atom 'execute_script' with 'args=["return {width: window.innerWidth, height: window.innerHeight}"]; frames='
[debug] [RemoteDebugger] Executing 'execute_script' atom in default context
[debug] [RemoteDebugger] Sending javascript command: '(function(){return (function(){var e=this||self...'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 88): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 88): '{"status":0,"value":{"width":0,"height":0}}'
[debug] [RemoteDebugger] Sending to Web Inspector took 47ms
[debug] [RemoteDebugger] Received result for atom 'execute_script' execution: {"width":0,"height":0}
[XCUITestDriver@e03c (311ea990)] Proxying to WDA with an unknown route: POST /wda/tap/0
[debug] [XCUITestDriver@e03c (311ea990)] Proxying [POST /wda/tap/0] to [POST http://192.168.4.255:8200/session/3E4DA667-C4B8-4653-B76A-602337458484/wda/tap/0] with body: {"x":0,"y":0}
[debug] [XCUITestDriver@e03c (311ea990)] Got response with status 200: {"value":null,"sessionId":"3E4DA667-C4B8-4653-B76A-602337458484"}
[debug] [RemoteDebugger] Sending javascript command: 'window.document.title'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2150', page '6', target 'page-87' (id: 90): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 90): '""'
[debug] [RemoteDebugger] Sending to Web Inspector took 31ms
[debug] [XCUITestDriver@e03c (311ea990)] undefined
[debug] [XCUITestDriver@e03c (311ea990)] Encountered internal error running command: Error: Cannot determine web view coordinates offset. Are you in Safari context?
[debug] [XCUITestDriver@e03c (311ea990)]     at XCUITestDriver.mobileCalibrateWebToRealCoordinatesTranslation (/Users/kazu/GitHub/appium-xcuitest-driver/lib/commands/web.js:890:13)
[debug] [XCUITestDriver@e03c (311ea990)]     at XCUITestDriver.executeMethod (/Users/kazu/GitHub/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/basedriver/commands/execute.ts:36:12)
[debug] [XCUITestDriver@e03c (311ea990)]     at XCUITestDriver.execute (/Users/kazu/GitHub/appium-xcuitest-driver/lib/commands/execute.js:118:14)
[HTTP] <-- POST /session/311ea990-ff4d-407f-a69b-5afac9c165e8/execute/sync 500 2528 ms - 702
```

Reply:
^ was iPad

Reply:
I have fixed width/height calculation. Could you please confirm the calibration page is loaded properly in the browser and click coordinates are visible in the title?

Reply:
Rarely the `title` result was an empty string as not complete the page load.

Maybe we could add retry a few times?

```
[debug] [XCUITestDriver@acfc (00ab5ac9)] Got response with status 200: {"value":null,"sessionId":"9B36A326-5333-4319-893C-660D05B69B75"}
[debug] [RemoteDebugger] Sending javascript command: 'window.document.title'
[debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:2226', page '4', target 'page-37' (id: 82): 'Runtime.evaluate'
[debug] [RemoteDebugger] Received data response from send (id: 82): '""'
[debug] [RemoteDebugger] Sending to Web Inspector took 44ms
[debug] [XCUITestDriver@acfc (00ab5ac9)] ""
[debug] [XCUITestDriver@acfc (00ab5ac9)] Encountered internal error running command: Error: Cannot determine web view coordinates offset. Are you in Safari context? Original error: Unexpected end of JSON input
[debug] [XCUITestDriver@acfc (00ab5ac9)]     at XCUITestDriver.mobileCalibrateWebToRealCoordinatesTranslation (/Users/kazu/GitHub/appium-xcuitest-driver/lib/commands/web.js:886:13)
[debug] [XCUITestDriver@acfc (00ab5ac9)]     at XCUITestDriver.executeMethod (/Users/kazu/GitHub/appium-xcuitest-driver/node_modules/@appium/base-driver/lib/basedriver/commands/execute.ts:36:12)
[debug] [XCUITestDriver@acfc (00ab5ac9)]     at XCUITestDriver.execute (/Users/kazu/GitHub/appium-xcuitest-driver/lib/commands/execute.js:118:14)
[HTTP] <-- POST /session/00ab5ac9-dcbd-42a4-be68-bbb3002b7eff/execute/sync 500 2832 ms - 792
```

Reply:
I also got negative offset and null value as title cases as below. This was only a simulator so far.

https://gist.github.com/KazuCocoa/a92ec14e1702edd8f9626d5c2405a6d1

Reply:
this might be happening because of scaling. Applied a separate condition for that case